app/models/note.rb

```diff
@@ -39,6 +49,10 @@ class Note < ApplicationRecord
 
   DEFAULT_FRESHLY_CLOSED_LIMIT = 7.days
 
+  def comments_with_extra_open_comment
```

Comment:
> Is this required in the model?

I think its needed for the behaviour of the API NotesController.

> […] and then note.comments which would return either N (if body migrated) or N-1 (if body not migrated) comments

I understand you are suggesting to implement a custom `Note#comments`-method instead of having the dedicated `Note#comments_with_extra_open_comment`-method?

I tried this out but realised quickly some issues with this:
a) We have calls to `Note#comments` from the (Web) NotesController context, which expects no prepended `open`-comment.

b) `Note#comments` returns an Active Record Collection Proxy-object. Overriding the `Note#comments`-behaviour so that it occasionally returns an Array instead, may lead to weird bugs. E.g. calls like `note.comments.create` will fail with a `NoMethodError`.

> Does it leak abstractions too much?

I am not sure what you mean with that. The idea of `Note#comments_with_extra_open_comment` is to hide the logic inside the model and make it obvious to developers that this method returns something special.

I've now added a `Note#api_comments`-method which should always return the full list of note comments including the `open`-comment.

Reply:
> I am not sure what you mean with that. The idea of `Note#comments_with_extra_open_comment` is to hide the logic inside the model and make it obvious to developers that this method returns something special.
> 
> I've now added a `Note#api_comments`-method which should always return the full list of note comments including the `open`-comment

But the point I'm making is that this still leaks complexity out of the Note model. We should have:

* Note#author
* Note#body
* Note#comments

That's all that both the web view and the API need. The web view can use them directly. The API can use the first two to make the legacy-first-comment.

I can see that the web view has some things like `if @note.body_migrated?` but the view should not care about this.

Now your point about the collection proxy object is a good one, and one I need to think about. Perhaps we can do something like (completely untested):

```
def comments
  if body_migrated?
    raw_comments
  else
    raw_comments.offset(1)
  end
end
```
... to get a collection proxy either way.

Partly this depends on the data migration strategy - whether we expect to do one of the two following options:

* A migration that, for each note (and in a transaction) creates the body, author, and deletes the first comment
* Two migrations, the first that backfills the body and author, and the second that deletes all the first comments. But that's extremely risky (imho) since the second migration is must-work-first-time.

Final point - there's still references to "opened comment" in the PR, but it's the "first comment" that we're focussing on.

Reply:
> The API can use the first two to make the legacy-first-comment.

Ah ok. You'd rather have the logic in the API controller, instead of hidden in a Model-method, correct?

That's ok for me. I'll rewrite the PR to reflect this.

> Partly this depends on the data migration strategy - whether we expect to do one of the two following options […]

I understand you lean towards the latter option – backfilling data and removing the first comment in 1 atomic operation. That should simplify the code as well I think. I'll adjust the PR to match this behaviour.